README.md
# Random

A Tiny Random Utility Generator

[![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](https://opensource.org/licenses/MIT)

### Features

* Alea RNG
* Seedable
* Random
    * number
    * between
    * chance
    * itemFromArray
    * entryFromObject
    * keyFromObject
* Tested with Jest
* 2kb minified (<1kb minified + gzipped)

### Install

```
git clone https://github.com/RetroVX/random.git
```
Or download from Zip

#### Node & Common
```javascript
const Random = require('./path/to/random.js');

const seed = 'mySeed';
const random = new Random(seed);
```

#### Module
```javascript
import Random from "./path/to/random.mjs";

const seed = 'mySeed';
const random = new Random(seed);
```

#### UMD
```html
<script type="text/javascript" src="path/to/random.umd.js"></script>
```
```javascript
const seed = 'mySeed';
const random = new window.random(seed);

```

### Examples

```javascript
// generate a random number between 0 and 1.
// use this in place of Math.random()
random.gen(); -> 0.9401234467513859

// generate a random number with a max number (defaults to 100 if undefined)
random.number(100); -> 42

// generates a random number using a min and max number
random.between(1, 25); -> 17

// percent chance to happen
random.chance(50) -> true

// picks a random item from an array
random.itemFromArray([1, 2, 3, 4, 5]); -> 2

// picks a random key/value pair and outputs as an array
random.entryFromObject({prop1: 'hello', prop2: 'world'}); -> ['prop1', 'hello']

// picks a random key from an object
random.keyFromObject({prop1: 'hello', prop2: 'world'}); -> 'prop2'

// new seed
random.newGenerator(seed);
```

### Credit
Alea
 * Johannes BaagÃ¸e <baagoe@baagoe.com>, 2010

### Version 1.3.0


CHANGELOG.md
## Random CHANGELOG

### Version 1.3.0

* new method ```chance```
    * returns a boolean depending on the chance of the passed in number

### Version 1.2.0

* new method ```keyFromObject```
    * returns a random key from a passed in object
    * added test

### Version 1.1.0

* created and built tests using jest

test/random.test.js
const RandomLib = require('./../dist/random.js');

// thanks to seeding, we know the outcome for each method allowing easier tests
// if the seed changes, the tests will need to be updated
const random = new RandomLib('jestTests');

describe('Testing the Random class', () => {

    test('random.gen should be between 0 and 1', () => {
        const value = random.gen();

        expect(value).toBeGreaterThanOrEqual(0);
        expect(value).toBeLessThanOrEqual(1);

    });

    test('random.number should output a number less than max argument (100)', () => {
        const max = 100;
        // random.number already defaults to a max of 100 if undefined but this helps for debugging
        const value = random.number(max);

        expect(value).toBeGreaterThanOrEqual(0);
        expect(value).toBeLessThanOrEqual(max);
    });

    test('random.between should ouput a number between the min (1) and max (25) numbers', () => {

        const value = random.between(1, 25);

        expect(value).toBeGreaterThanOrEqual(1);
        expect(value).toBeLessThanOrEqual(25);
        expect(value).toBeLessThan(26);

    });

    test('random.chance should return a boolean', () => {

        const value = random.chance(50);
        expect(value).toBeFalsy();
    })

    test('random.itemFromArray should output a random item from the passed in array', () => {
        const mockArray = [1, 2, 3, 4, 5];
        // item should always be '3'
        const item = random.itemFromArray(mockArray);

        expect(item).toBe(3);
    });

    test('random.entryFromObject should output an array with a key value pair', () => {
        const testObject = {
            prop1: 'hello',
            prop2: 'world',
            prop3: 42,
        }
        const entry = random.entryFromObject(testObject);
        expect(Array.isArray(entry)).toBeTruthy();

        expect(entry).toContain('prop3');
        expect(entry).toContain(42);

        expect(entry[0]).toBe('prop3');
        expect(entry[1]).toBe(42);

        expect(testObject).toHaveProperty(entry[0], entry[1]);

    });

    test('random.keyFromObject should return a random key from an object', () => {
        const testObject = {
            prop1: 'hello',
            prop2: 'world',
            prop3: 42,
        }
        // should always be 'prop3'
        const key = random.keyFromObject(testObject);

        expect(key).toBe('prop3');
        expect(key).toMatch('prop3');
    })

});

src/alea.js
/**
 * Alea
 * From http://baagoe.com/en/RandomMusings/javascript/
 * Johannes BaagÃ¸e <baagoe@baagoe.com>, 2010
 */

function Mash() {
    var n = 0xefc8249d;

    var mash = function (data) {
        data = data.toString();
        for (var i = 0; i < data.length; i++) {
            n += data.charCodeAt(i);
            var h = 0.02519603282416938 * n;
            n = h >>> 0;
            h -= n;
            h *= n;
            n = h >>> 0;
            h -= n;
            n += h * 0x100000000; // 2^32
        }
        return (n >>> 0) * 2.3283064365386963e-10; // 2^-32
    };

    mash.version = 'Mash 0.9';
    return mash;
}

// From http://baagoe.com/en/RandomMusings/javascript/
export function Alea() {
    return (function (args) {
        // Johannes BaagÃ¸e <baagoe@baagoe.com>, 2010
        var s0 = 0;
        var s1 = 0;
        var s2 = 0;
        var c = 1;

        if (args.length == 0) {
            args = [+new Date()];
        }
        var mash = Mash();
        s0 = mash(' ');
        s1 = mash(' ');
        s2 = mash(' ');

        for (var i = 0; i < args.length; i++) {
            s0 -= mash(args[i]);
            if (s0 < 0) {
                s0 += 1;
            }
            s1 -= mash(args[i]);
            if (s1 < 0) {
                s1 += 1;
            }
            s2 -= mash(args[i]);
            if (s2 < 0) {
                s2 += 1;
            }
        }
        mash = null;

        var random = function () {
            var t = 2091639 * s0 + c * 2.3283064365386963e-10; // 2^-32
            s0 = s1;
            s1 = s2;
            return s2 = t - (c = t | 0);
        };
        random.uint32 = function () {
            return random() * 0x100000000; // 2^32
        };
        random.fract53 = function () {
            return random() +
                (random() * 0x200000 | 0) * 1.1102230246251565e-16; // 2^-53
        };
        random.version = 'Alea 0.9';
        random.args = args;
        return random;

    }(Array.prototype.slice.call(arguments)));
};


src/random.js
import { Alea } from "./alea.js";

/**
 * @author Conor Irwin <https://github.com/RetroVX> 
 * @license {@link http://opensource.org/licenses/MIT|MIT License}
 * @classdesc
 * A basic random utility generator using alea for seedable randomness  
 * Credit Alea: 
 * From http://baagoe.com/en/RandomMusings/javascript/
 * Johannes BaagÃ¸e <baagoe@baagoe.com>, 2010  
 * @class Random
 * @version 1.3.0
 * @example
 * const random = new Random('mySeed');
 * random.number(); 
 * -> 42
 * // instead of Math.random()
 * random.gen();
 * -> 0.9401234467513859
 */
export default class Random {

    constructor(seed) {

        /**
         * the generator that creates random numbers.
         * use in place of Math.random()
         * @type method
         * @name Random.gen
         * @example
         * random.gen() -> 0.9401234467513859
         */
        this.gen = null;


        /**
         * The seed for alea to use
         * @name Random.seed
         */
        this.seed = null;
        

        // Create the generator
        this.newGenerator(seed);

    }


    /**
     * Create a new Alea generator with a new seed
     * @method Random.newGenerator
     * @param {*} seed - a number or string to use as the seed
     */
    newGenerator(seed) {        
        this.seed = seed || new Date().getTime();
        this.gen = new Alea(this.seed);
    }


    /**
     * Return a random number depending on the max number inputted. 
     * If the max argument is undefined, max = 100.
     * @param {number} max - the max number to be able to generate
     * @returns {number}
     * @example
     * random.number(); -> 56
     * random.number(50000) -> 3741
     */
    number(max) {
        if(max === undefined || max === null) { max = 100; }

        return Math.floor(this.gen() * Math.floor(max));
    }


    /**
     * Returns a number that is between the min and max arguments passed in
     * @method Random.between
     * @param {number} min - the min number
     * @param {number} max - the max number
     * @returns {number}
     * @example 
     * random.between(1, 25) -> 17
     */
    between(min, max) { 
        return Math.floor(this.gen() * (max - min + 1)) + min;
    }


    /**
     * Returns a boolean depending on if the percentChance check is passed
     * @param {number} percentChance - the number (percent out of 100)
     * @return {boolean}
     * @example
     * random.chance(50) -> 50% chance of happening
     * random.chance(1) -> 1% chance of happening
     */
    chance(percentChance) {
        return !!percentChance && this.gen() * 100 <= percentChance;
    }


    /**
     * Returns a random item from an array
     * @method Random.itemFromArray
     * @param {array} userArray - the array to pass in
     * @returns {*} item from array
     * @example
     * const list = [1, 2, 3, 4, 5];
     * const randomFromList = random.itemFromArray(list);
     * -> 2
     * const list2 = [{key: 'hello', something: 'world'}, {key: 'world', something: 'yes'}];
     * const randomFromList2 = random.itemFromArray(list2);
     * -> {key: 'hello', something: 'world'}
     */
    itemFromArray(userArray) {
        return userArray[Math.floor(this.gen() * userArray.length)];
    }


    /**
     * Returns a random key/value pair from an object
     * @method Random.entryFromObject
     * @param {object} userObject - the object to pass in
     * @returns {array} Returns an array with the random key and value pair
     * @example
     * const myObject = {
     *     prop: 'hello',
     *     prop2: 'world',
     *     prop3: 42
     * }
     * const randomFromObject = random.entryFromObject(myObject);
     * -> ['prop2', 'world']
     */
    entryFromObject(userObject) {
        const entries = Object.entries(userObject);
        const randomProp = this.itemFromArray(entries);

        return randomProp;
    }


    /**
     * Returns a random key from an object
     * @method Random.keyFromObject
     * @param {object} userObject - the object to pass in
     * @returns {string} Returns a string with the random key
     * @example
     * const myObject = {
     *     prop: 'hello',
     *     prop2: 'world',
     *     prop3: 42
     * }
     * const randomKeyFromObject = random.keyFromObject(myObject);
     * -> 'prop3'
     */
    keyFromObject(userObject) {
        const keys = Object.keys(userObject);
        const randomKey = this.itemFromArray(keys);

        return randomKey;
    }
}